Vetka2 sql

andr777 opened the pull request:
все работает но почемуто 500 пишет почему не пойму

andr777 committed:
1

andr777 committed:
sql

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @PutMapping
+    public ResponseEntity<Faculty> editFaculty(@RequestBody long id, Faculty faculty) {
```
здесь неверно. Аннотация RequestBody должна быть на факультете, а id Должен быть в пути

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,60 @@
+    public ResponseEntity<Faculty> editFaculty(@RequestBody long id, Faculty faculty) {
+        Faculty foundFaculty = facultyService.editFaculty(id, faculty);
+        if (foundFaculty == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
not found

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,60 @@
+
+    // Controller
+    @GetMapping("/{color}")
+    public Collection<Faculty> findByColor(@RequestParam(required = false) String color) {
```
почему required=false? должно быть true

dmitry-bizin replied:
убери required

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,58 @@
+        return studentService.findStudent(id);
+    }
+    @PutMapping
+    public ResponseEntity<Student> editStudent(@RequestBody  long id, Student student) {
```
здесь неверно. Аннотация RequestBody должна быть на студенте, а id Должен быть в пути

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,58 @@
+    public ResponseEntity<Student> editStudent(@RequestBody  long id, Student student) {
+        Student foundStudent = studentService.editStudent(id, student);
+        if (foundStudent == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
not found

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,58 @@
+
+    // Controller
+    @GetMapping("/{age}")
+    public Collection<Student> findStudents(@RequestParam(required = false) int age) {
```
почему false-то? должно быть true. И условие никакое не нужно в методе

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Faculty findFaculty(long id) {
+        return facultyRepository.findById(id).get();
```
вот поэтому у тебя и 500, потому что здесь исключение. Я уже 100 раз говорил на разборах, что get не надо никогда вызывать без проверки. Он просто выкинет исключение, если не будет элемента. У тебя контроллер выше ожидаешь налл, поэтому здесь надо вместо get orElse(null)

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/FacultyService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Faculty editFaculty(long id, Faculty faculty) {
+        return facultyRepository.save(faculty);
```
а где проверка, что такая сущность есть? будет просто сохранение, а надо вернуть налл, если не будет сущности по id

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Student findStudent(long id) {
+        return studentRepository.findById(id).get();
```
вот поэтому у тебя и 500, потому что здесь исключение. Я уже 100 раз говорил на разборах, что get не надо никогда вызывать без проверки. Он просто выкинет исключение, если не будет элемента. У тебя контроллер выше ожидаешь налл, поэтому здесь надо вместо get orElse(null)

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Student editStudent(long id, Student student) {
+        return studentRepository.save(student);
```
а где проверка, что такая сущность есть? будет просто сохранение, а надо вернуть налл, если не будет сущности по id

andr777 committed:
????

andr777 committed:
вроде работает

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -21,7 +21,7 @@ public FacultyController(FacultyService facultyService) {
     }
 
-    @GetMapping("/{id}")
+    @GetMapping("/getid/{id}")
```
не нужно getid в пути

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -30,12 +30,12 @@ public ResponseEntity<Faculty> getFacultyInfo(@PathVariable long id) {
     }
 
-    @PostMapping
+    @PostMapping("/add")
```
не нужно add, это тавтология. POST уже говорит о том, что добавление

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -30,12 +30,12 @@ public ResponseEntity<Faculty> getFacultyInfo(@PathVariable long id) {
     }
 
-    @PutMapping
+    @PutMapping("/edit/{id}")
```
не нужно edit, это тавтология. PUTуже говорит о том, что обновление

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -44,14 +44,14 @@ public ResponseEntity<Faculty> editFaculty(@PathVariable long id, @RequestBody F
     }
 
-    @DeleteMapping("/{id}")
+    @DeleteMapping("/delid/{id}")
```
не нужно delid, это тавтология. DELETEуже говорит о том, что удаление

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -44,14 +44,14 @@ public ResponseEntity<Faculty> editFaculty(@PathVariable long id, @RequestBody F
 
     // Controller
-    @GetMapping("/{color}")
+    @GetMapping("/getcolor/{color}")
```
не нужно getcolor

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -20,20 +20,20 @@ public StudentController(StudentService studentService) {
     }
 
-    @PostMapping
+    @PostMapping("/add")
```
не нужно add, это тавтология. POST уже говорит о том, что добавление

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -20,20 +20,20 @@ public StudentController(StudentService studentService) {
         return ResponseEntity.ok(foundStudent);
     }
-    @PutMapping
+    @PutMapping("/edit/{id}")
```
не нужно edit, это тавтология. PUTуже говорит о том, что обновление

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -20,20 +20,20 @@ public StudentController(StudentService studentService) {
     }
 
-    @GetMapping("/{id}")
+    @GetMapping("/getid/{id}")
```
не нужно getid в пути

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -42,22 +42,21 @@ public ResponseEntity<Student> editStudent(@PathVariable long id, @RequestBody S
     }
 
-    @DeleteMapping("/{id}")
+    @DeleteMapping("/delid/{id}")
```
не нужно delid, это тавтология. DELETEуже говорит о том, что удаление

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -42,22 +42,21 @@ public ResponseEntity<Student> editStudent(@PathVariable long id, @RequestBody S
 
     // Controller
-    @GetMapping("/{age}")
+    @GetMapping("/getage/{age}")
```
не нужно getage

dmitry-bizin commented on src/main/java/ru/hogwarts/school/service/StudentService.java:
```diff
@@ -0,0 +1,57 @@
+
+    // Service
+    public Collection<Student> findByAge(int age) {
+        List<Student> sortedStudents = getAll().stream()
```
здесь надо использовать метод репозитория, а не в коде фильтровать

andr777 committed:
метод поиска студентов по возрастам фильтрует в коде, а не на уровне БД.
переделал как в поиске по цвету факультета.

в этом методе стало выдавать 500. ничего не ищет.

зачем-то ты поломал REST: написал в путях слова вида add, edit, delid, ... зачем? Это лишнее и нарушает принципы REST, нужно вернуть, как было.

убрал в путях "add, edit, delid" итд.

стало выдавать 500 и 404 по всему остальному.

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    // Controller
+    @GetMapping("/{color}")
```
здесь не нужно в пути ничего писать, просто `@GetMapping`

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    // Controller
+    @GetMapping("/{age}")
```
здесь тоже `@GetMapping` и всё

dmitry-bizin commented on src/main/java/ru/hogwarts/school/controller/StudentController.java:
```diff
@@ -0,0 +1,66 @@
+
+    // Controller
+    @GetMapping("/{age}")
+    public Collection<Student> findStudents(@PathVariable int age) {
```
Не PathVariable, а RequestParam

andr777 committed:
ok
